// videogular2/src/controls/vg-scrub-bar/vg-scrub-bar.js
"use strict";
var core_1 = require('@angular/core');
var vg_api_1 = require('../../core/services/vg-api');
var VgScrubBar = (function () {
    function VgScrubBar(ref, API) {
        this.API = API;
        this.elem = ref.nativeElement;
    }
    VgScrubBar.prototype.ngOnInit = function () {
        var _this = this;
        this.API.playerReadyEvent.subscribe(function () { return _this.onPlayerReady(); });
    };
    VgScrubBar.prototype.onPlayerReady = function () {
        this.target = this.API.getMediaById(this.vgFor);
    };
    VgScrubBar.prototype.onMouseDownScrubBar = function ($event) {
        if (!this.target.isLive) {
            var percentage = $event.offsetX * 100 / this.elem.scrollWidth;
            this.target.seekTime(percentage, true);
        }
    };
    VgScrubBar.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'vg-scrub-bar',
                    encapsulation: core_1.ViewEncapsulation.None,
                    template: "<ng-content></ng-content>",
                    styles: ["\n        vg-scrub-bar {\n            position: absolute;\n            width: 100%;\n            height: 5px;\n            bottom: 50px;\n            margin: 0;\n            cursor: pointer;\n            align-items: center;\n            background: rgba(0, 0, 0, 0.75);\n            z-index: 250;\n        }\n\n        vg-controls vg-scrub-bar {\n            position: relative;\n            bottom: initial;\n            background: initial;\n            height: 50px;\n            flex-grow: 1;\n            flex-basis: 0;\n            margin: 0 10px;\n        }\n    "]
                },] },
    ];
    /** @nocollapse */
    VgScrubBar.ctorParameters = [
        { type: core_1.ElementRef, },
        { type: vg_api_1.VgAPI, },
    ];
    VgScrubBar.propDecorators = {
        'vgFor': [{ type: core_1.Input },],
        'onMouseDownScrubBar': [{ type: core_1.HostListener, args: ['mousedown', ['$event'],] },],
    };
    return VgScrubBar;
}());
exports.VgScrubBar = VgScrubBar;
