// videogular2/src/controls/vg-controls.js
"use strict";
var core_1 = require('@angular/core');
var Observable_1 = require("rxjs/Observable");
var vg_api_1 = require("../core/services/vg-api");
require('rxjs/add/observable/fromEvent');
var VgControls = (function () {
    function VgControls(API, ref) {
        this.API = API;
        this.ref = ref;
        this.isAdsPlaying = 'initial';
        this.hideControls = false;
        this.vgAutohide = false;
        this.vgAutohideTime = 3;
        this.elem = ref.nativeElement;
    }
    VgControls.prototype.ngOnInit = function () {
        var _this = this;
        var mouseEnter = Observable_1.Observable.fromEvent(this.API.videogularElement, 'mouseenter');
        mouseEnter.subscribe(this.show.bind(this));
        var mouseLeave = Observable_1.Observable.fromEvent(this.API.videogularElement, 'mouseleave');
        mouseLeave.subscribe(this.hide.bind(this));
        this.API.playerReadyEvent.subscribe(function () { return _this.onPlayerReady(); });
    };
    VgControls.prototype.onPlayerReady = function () {
        this.target = this.API.getMediaById(this.vgFor);
        this.target.subscriptions.startAds.subscribe(this.onStartAds.bind(this));
        this.target.subscriptions.endAds.subscribe(this.onEndAds.bind(this));
    };
    VgControls.prototype.ngAfterViewInit = function () {
        if (this.vgAutohide) {
            this.hide();
        }
        else {
            this.show();
        }
    };
    VgControls.prototype.onStartAds = function () {
        this.isAdsPlaying = 'none';
    };
    VgControls.prototype.onEndAds = function () {
        this.isAdsPlaying = 'initial';
    };
    VgControls.prototype.hide = function () {
        if (this.vgAutohide) {
            clearTimeout(this.timer);
            this.hideAsync();
        }
    };
    VgControls.prototype.show = function () {
        clearTimeout(this.timer);
        this.hideControls = false;
    };
    VgControls.prototype.hideAsync = function () {
        var _this = this;
        this.timer = setTimeout(function () {
            _this.hideControls = true;
        }, this.vgAutohideTime * 1000);
    };
    VgControls.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'vg-controls',
                    encapsulation: core_1.ViewEncapsulation.None,
                    template: "<ng-content></ng-content>",
                    styles: ["\n        vg-controls {\n            position: absolute;\n            display: flex;\n            width: 100%;\n            height: 50px;\n            z-index: 300;\n            bottom: 0;\n            background-color: rgba(0, 0, 0, 0.5);\n            -webkit-transition: bottom 1s;\n            -khtml-transition: bottom 1s;\n            -moz-transition: bottom 1s;\n            -ms-transition: bottom 1s;\n            transition: bottom 1s;\n        }\n\n        vg-controls.hide {\n          bottom: -50px;\n        }\n    "]
                },] },
    ];
    /** @nocollapse */
    VgControls.ctorParameters = [
        { type: vg_api_1.VgAPI, },
        { type: core_1.ElementRef, },
    ];
    VgControls.propDecorators = {
        'isAdsPlaying': [{ type: core_1.HostBinding, args: ['style.pointer-events',] },],
        'hideControls': [{ type: core_1.HostBinding, args: ['class.hide',] },],
        'vgAutohide': [{ type: core_1.Input },],
        'vgAutohideTime': [{ type: core_1.Input },],
    };
    return VgControls;
}());
exports.VgControls = VgControls;
